### PR TITLE
Make ParquetFileWriter use TrinoOutputFile

### PR DESCRIPTION
**Pull request by homar**
## Description
This is prerequisite to be able to track FileSystem memory utilization while writing Parquet data - https://github.com/trinodb/trino/issues/14023

## Non-technical explanation
Allows better tracking of memory consumed internally.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 1]
```diff
@@ -176,14 +177,13 @@ private List<Slice> rewriteFile(Path sourcePath, FileDeletion deletion)
...
-            FileSystem fileSystem = hdfsEnvironment.getFileSystem(new HdfsContext(session.getIdentity()), rootTablePath);
 
             Path targetPath = new Path(sourcePath.getParent(), session.getQueryId() + "_" + randomUUID());
             String targetRelativePath = rootTablePath.toUri().relativize(targetPath.toUri()).toString();
-            FileWriter fileWriter = createParquetFileWriter(fileSystem, targetPath, dataColumns);
+            FileWriter fileWriter = createParquetFileWriter(targetPath.toString(), dataColumns);
 
             DeltaLakeWriter writer = new DeltaLakeWriter(
-                    fileSystem,
+                    hdfsEnvironment.getFileSystem(new HdfsContext(session.getIdentity()), rootTablePath),
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:186]: Do we still need `hdfsEnvironment`?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 2]
```diff
@@ -201,16 +201,17 @@ private List<Slice> rewriteFile(Path sourcePath, FileDeletion deletion)
...
 
-    private FileWriter createParquetFileWriter(FileSystem fileSystem, Path path, List<DeltaLakeColumnHandle> dataColumns)
+    private FileWriter createParquetFileWriter(String path, List<DeltaLakeColumnHandle> dataColumns)
     {
         ParquetWriterOptions parquetWriterOptions = ParquetWriterOptions.builder()
                 .setMaxBlockSize(getParquetWriterBlockSize(session))
                 .setMaxPageSize(getParquetWriterPageSize(session))
                 .build();
         CompressionCodecName compressionCodecName = getCompressionCodec(session).getParquetCompressionCodec();
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:211]: Rather than creating file system over again, please put it in `DeltaLakeMergeSink` as a field instead of `fileSystemFactory`.

The same applies to other classes like `DeltaLakeUpdatablePageSource`, `HiveWriterFactory`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetFileWriter.java [hunk 3]
```diff
@@ -62,7 +63,7 @@
     private long validationCpuNanos;
 
     public ParquetFileWriter(
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetFileWriter.java:65]: add throws `IOException` rather than using `UncheckedIOException`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 1 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:186]: It is not that simple as it requires using `trino-parquet` MetadataReader instead of hive's one but I will give it a try

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java [hunk 4]
```diff
@@ -100,6 +101,7 @@
 
     private final PageIndexer pageIndexer;
     private final HdfsEnvironment hdfsEnvironment;
+    private final TrinoFileSystem fileSystemFactory;
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:104]: `fileSystemFactory` -> `fileSystem`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java [hunk 5]
```diff
@@ -100,6 +101,7 @@
 
     private final PageIndexer pageIndexer;
     private final HdfsEnvironment hdfsEnvironment;
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:103]: Do we still need this guy?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java [hunk 6]
```diff
@@ -409,7 +413,7 @@ private int[] getWriterIndexes(Page page)
                 writers.set(writerIndex, writer);
                 memoryUsage += writer.getMemoryUsage();
             }
-            catch (IOException e) {
+            catch (RuntimeException e) {
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:416]: Can we still use `IOException` here?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java [hunk 7]
```diff
@@ -189,16 +193,21 @@ private static DeltaLakeJsonFileStatistics readStatistics(
             typeForColumn.put(dataColumnNames.get(i), dataColumnTypes.get(i));
         }
 
+        TrinoInputFile inputFile = fileSystem.newInputFile(new Path(tableLocation, relativeFilePath).toString());
+        TrinoParquetDataSource trinoParquetDataSource = new TrinoParquetDataSource(
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:197]: wrap in try with resources, otherwise it might remain unclosed

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveWriterFactory.java [hunk 8]
```diff
@@ -589,7 +588,6 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
             else {
                 tempFilePath = new Path(path.getParent(), ".tmp-sort." + path.getName());
             }
-            fileSystem = fileSystemFactory.create(session);
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveWriterFactory.java:592]: nit: could be separate commit

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/HdfsParquetDataSource.java [hunk 9]
```diff
@@ -30,26 +29,26 @@
...
 {
-    private final FSDataInputStream inputStream;
+    private final TrinoInput input;
     private final FileFormatDataSourceStats stats;
 
     public HdfsParquetDataSource(
             ParquetDataSourceId id,
             long estimatedSize,
-            FSDataInputStream inputStream,
+            TrinoInput input,
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/HdfsParquetDataSource.java:38]: Can it be `io.trino.filesystem.TrinoInputFile` so that `HdfsParquetDataSource` manages input lifecycle?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java (hunk 5 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:103]: not as a class field

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java (hunk 6 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:416]: No we can't it is no longer thrown here. Previously it was thrown by `hdfsEnvironment.getFileSystem(session.getIdentity(), rootTableLocation, conf)`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java [hunk 10]
```diff
@@ -224,7 +226,7 @@ public long getValidationCpuNanos()
     @Override
     public CompletableFuture<Collection<Slice>> finish()
     {
-        ListenableFuture<Collection<Slice>> result = hdfsEnvironment.doAs(session.getIdentity(), this::doFinish);
+        ListenableFuture<Collection<Slice>> result = doFinish();
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:229]: inline `doFinish` method

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java [hunk 11]
```diff
@@ -276,7 +273,7 @@ private void doAbort()
     public CompletableFuture<?> appendPage(Page page)
     {
         if (page.getPositionCount() > 0) {
-            hdfsEnvironment.doAs(session.getIdentity(), () -> doAppend(page));
+            doAppend(page);
```
Comment (sopel39) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSink.java:276]: inline:
```
        if (page.getPositionCount() == 0) {
            return NOT_BLOCKED;
        }

        while (page.getPositionCount() > MAX_PAGE_POSITIONS) {
            Page chunk = page.getRegion(0, MAX_PAGE_POSITIONS);
            page = page.getRegion(MAX_PAGE_POSITIONS, page.getPositionCount() - MAX_PAGE_POSITIONS);
            writePage(chunk);
        }

        writePage(page);
        return NOT_BLOCKED;
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/HdfsParquetDataSource.java [hunk 12]
```diff
@@ -30,26 +30,27 @@
 public class HdfsParquetDataSource
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/HdfsParquetDataSource.java:30]: This class should be deleted

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveWriterFactory.java [hunk 13]
```diff
@@ -200,7 +200,7 @@ public HiveWriterFactory(
             HiveWriterStats hiveWriterStats)
     {
         this.fileWriterFactories = ImmutableSet.copyOf(requireNonNull(fileWriterFactories, "fileWriterFactories is null"));
-        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        this.fileSystem = requireNonNull(fileSystemFactory, "fileSystemFactory is null").create(session);
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveWriterFactory.java:203]: `requireNonNull` is not needed when we call a method on an object

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestHdfsParquetDataSource.java [hunk 14]
```diff
@@ -20,36 +20,43 @@
...
 {
     @Test(dataProvider = "testPlanReadOrderingProvider")
     public void testPlanReadOrdering(DataSize maxBufferSize)
+            throws IOException
     {
         Slice testingInput = Slices.wrappedIntArray(IntStream.range(0, 1000).toArray());
-        HdfsParquetDataSource dataSource = new HdfsParquetDataSource(
-                new ParquetDataSourceId("test"),
-                0,
-                new FSDataInputStream(new TestingSliceInputStream(testingInput.getInput())),
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestHdfsParquetDataSource.java:50]: TestingSliceInputStream seems to be unused now, it can be deleted

Comment (homar) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestHdfsParquetDataSource.java:50]: Good point, I haven't noticed that

**Review by raunaqmorarka — APPROVED** [review]

**Review by sopel39 — APPROVED** [review]
